File: trail4/N개 중에 M개 뽑기/n-choose-m.cpp
#include <vector>
#include <iostream>

using namespace std;

int N, M;

void comb(int start, vector<int> arr, vector<bool> chk){
    if(arr.size() == M){
        for(int val : arr){
            cout << val << " ";
        }
        cout << endl;
        return;
    }

    for(int i = start; i<=N; i++){
        if(chk[i] == true) continue;
        arr.push_back(i);
        chk[i] = true;
        comb(i+1, arr, chk);
        arr.pop_back();
        chk[i] = false;
    }
}

int main() {
    cin >> N >> M;
    vector<int> v;
    vector<bool> chk(N+1, false);
    comb(1, v, chk);

    return 0;
}

File: trail4/N개 중에 M개 뽑기/n-choose-m.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, m;
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();
        comb(1, new int[m], 0, 0);
        

        bw.flush();
        bw.close();
    }
    public static void comb(int start, int[] arr, int idx, int mask) throws IOException {
        if(idx == m){
            for(int v : arr){
                bw.write(v+" ");
            }
            bw.newLine();
            return;
        }

        for(int i = start; i<=n; i++){
            if((mask & (1<<i)) != 0) continue;
            mask |= (1<<i);
            arr[idx] = i;
            comb(i+1, arr, idx+1, mask);
            mask &= ~(1<<i);
        }
    }
}